modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,403 @@
+      // `req`, `message` and `data` are optional.
+      // `req` is an apos request object. If provided, the `data` argument will be
+      // enriched with additional information from the request.
+      // `data.module` is recognized as a special property and is used to refortmat
+      // the message.
+      // `data.severity` is always set to the value of the `severity` argument.
+      // `data.type` is always set to the value of the `eventType` argument.
+      // message is optional, if not provided it is generated from the eventType and data.module.
+      // Optional message values:
+      // - `module: eventType: message`
+      // - `eventType: message`: (missing module)
+      // - `module: eventType`: (missing message)
+      // - `eventType`: (missing module and message)
+      //
+      // Returns [ data ] or [ message, data ] depending on option.messageAs.
+      // `data` is always an object containing at least a `type` and `severity` properties.
+      processLoggerArgs(moduleSelf, ...args) {
+        let severity;
+        let req;
+        let eventType;
+        let message;
+        let data;
+
+        // Detect `req` argument with a simple duck type check - apos `req` object
+        // has always a translate `t` function.
+        if (args[1] && typeof args[1].t === 'function') {
+          [ severity, req, eventType, message, data ] = args;
+        } else {
+          [ severity, eventType, message, data ] = args;
+        }
+
+        if (typeof eventType !== 'string') {
+          throw new Error('Event type must be a string');
+        }
+        if (_.isPlainObject(message)) {
+          data = message;
+          message = undefined;
+        }
+        data = data ? { ...data } : {};
+        const aposModule = moduleSelf.__meta?.name ?? '__unknown__';
+
+        if (typeof message === 'string' && message.trim().length > 0) {
+          message = aposModule
+            ? `${aposModule}: ${eventType}: ${message}`
+            : `${eventType}: ${message}`;
+        } else {
+          message = aposModule
+            ? `${aposModule}: ${eventType}`
+            : eventType;
+        }
+        if (self.options.messageAs) {
+          data[self.options.messageAs] = message;
+        }
+        data.module = aposModule;
+        data.type = eventType;
+        data.severity = severity;
+
+        data = self.processRequestData(req, data);
+
+        return [ message, data ];
+      },
+
+      // Enrich the `data` argument with additional information from the request.
+      processRequestData(req, data) {
+        if (!req) {
+          return data;
+        }
+
+        // Keep the proper property order for better readability.
+        const message = {};
+        if (self.options.messageAs) {
+          message[self.options.messageAs] = data[self.options.messageAs];
+        }
+        const {
+          module: _module, type, severity, ...rest
+        } = data;
+
+        return {
```

Comment:
Don't clone the object again, we know it is already a clone and it hurts performance. Assign the new properties to the existing object.

Reply:
Yeah, last moment change because of the prop order requirement. But there is an easy way to be done via a mutation!